torchmdnet/datasets/custom.py

```diff
@@ -30,87 +60,161 @@ class Custom(Dataset):
         self.has_energies = energyglob is not None
         self.has_forces = forceglob is not None
-
-        self.coordfiles = sorted(glob.glob(coordglob))
-        self.embedfiles = sorted(glob.glob(embedglob))
-        self.energyfiles = sorted(glob.glob(energyglob)) if self.has_energies else None
-        self.forcefiles = sorted(glob.glob(forceglob)) if self.has_forces else None
-
-        assert len(self.coordfiles) == len(self.embedfiles), (
-            f"Number of coordinate files {len(self.coordfiles)} "
-            f"does not match number of embed files {len(self.embedfiles)}."
+        self.fields = [
+            ("pos", "pos", torch.float32),
+            ("z", "types", torch.long),
+        ]
+        self.files = {}
+        self.files["pos"] = sorted(glob.glob(coordglob))
+        self.files["z"] = sorted(glob.glob(embedglob))
+        assert len(self.files["pos"]) == len(self.files["z"]), (
+            f"Number of coordinate files {len(self.files['pos'])} "
+            f"does not match number of embed files {len(self.files['z'])}."
         )
         if self.has_energies:
-            assert len(self.coordfiles) == len(self.energyfiles), (
-                f"Number of coordinate files {len(self.coordfiles)} "
-                f"does not match number of energy files {len(self.energyfiles)}."
+            self.files["y"] = sorted(glob.glob(energyglob))
+            self.fields.append(("y", "energy", torch.float32))
+            assert len(self.files["pos"]) == len(self.files["y"]), (
+                f"Number of coordinate files {len(self.files['pos'])} "
+                f"does not match number of energy files {len(self.files['y'])}."
             )
         if self.has_forces:
-            assert len(self.coordfiles) == len(self.forcefiles), (
-                f"Number of coordinate files {len(self.coordfiles)} "
-                f"does not match number of force files {len(self.forcefiles)}."
+            self.files["neg_dy"] = sorted(glob.glob(forceglob))
+            self.fields.append(("neg_dy", "forces", torch.float32))
+            assert len(self.files["pos"]) == len(self.files["neg_dy"]), (
+                f"Number of coordinate files {len(self.files['pos'])} "
+                f"does not match number of force files {len(self.files['neg_dy'])}."
             )
-
-        print("Number of files: ", len(self.coordfiles))
-
+        print("Number of files: ", len(self.files["pos"]))
+        self.cached = False
+        if read_as_hdf5 is not None:
+            hdf5_file = read_as_hdf5
+            write_as_hdf5(
+                self.files,
+                hdf5_file,
+            )
+            self.hdf5_dataset = HDF5(hdf5_file)
+        total_data_size = self._initialize_index()
+        print("Combined dataset size {}".format(len(self.index)))
+        # If the dataset is small enough, load it whole into CPU memory
+        data_size_limit = preload_memory_limit * 1024 * 1024
+        self.cache = False
```

Comment:
Check this, I think you meant `self.cached`  but you already defined it in line 104.